angel_system/global_step_prediction/global_step_predictor.py

```diff
@@ -148,70 +159,164 @@ def initialize_new_recipe_tracker(self, recipe):
+                    "full_str": "background",
                 },
             )
+        print(recipe)
+        tracker_dict[
+            "first_granular_step_per_broad_step"
+        ] = self.get_first_granular_step_per_broad_step(broad_steps)
         tracker_dict["recipe"] = recipe
-        tracker_dict["current_step"] = 0
-        tracker_dict["total_num_steps"] = len(steps)
-        tracker_dict["can_skip"] = False
-        tracker_dict["step_to_activity_id"] = [step["activity_id"] for step in steps]
-        tracker_dict["step_to_activity_desc"] = [step["description"] for step in steps]
-        tracker_dict["prediction_history"] = np.array([])
+        tracker_dict["current_broad_step"] = 0
+        tracker_dict["current_granular_step"] = 0
+        tracker_dict["total_num_broad_steps"] = len(broad_steps)
+        tracker_dict["total_num_granular_steps"] = np.sum(
+            [len(self.get_unique(step["activity_ids"])) for step in config["labels"]]
+        )
+        tracker_dict["broad_step_to_activity_ids"] = [
+            self.get_unique(step["activity_ids"]) for step in broad_steps
+        ]
+        tracker_dict[
+            "granular_step_to_activity_id"
+        ] = self.get_activity_per_granular_step(broad_steps)
+        tracker_dict["step_to_label"] = [step["label"] for step in broad_steps]
+        tracker_dict["step_to_full_str"] = [step["full_str"] for step in broad_steps]
+        tracker_dict["broad_step_prediction_history"] = np.array([])
+        tracker_dict["granular_step_prediction_history"] = np.array([])
         tracker_dict["active"] = True
-        tracker_dict["steps"] = steps
+        tracker_dict["broad_steps"] = broad_steps
+        tracker_dict["can_skip"] = True
 
         self.trackers.append(tracker_dict)
 
+    def get_activity_per_granular_step(self, broad_steps):
+        activity_id_per_granular_step = []
+        for broad_step in broad_steps:
+            for activity_id in self.get_unique(broad_step["activity_ids"]):
+                activity_id_per_granular_step.append(activity_id)
+        return activity_id_per_granular_step
+
+    def increment_granular_step(self, tracker_ind):
+        """
+        Increment a tracker's granular step, and also update the tracker's
+        broad step.
+        (Many broad steps contain several granular steps, so a granular step
+        increment may or may not entail a broad step increment.)
+        """
+        tracker = self.trackers[tracker_ind]
+        if tracker["current_granular_step"] < tracker["total_num_granular_steps"] - 1:
+            self.trackers[tracker_ind]["current_granular_step"] += 1
+            self.trackers[tracker_ind][
+                "current_broad_step"
+            ] = self.granular_to_broad_step(tracker, tracker["current_granular_step"])
+        else:
+            raise Exception(
+                f"Tried to increment tracker #{tracker_ind}: \
+                    {tracker['recipe']} past last step."
```

Comment:
String continuation that should be using separate strings
```suggestion
                f"Tried to increment tracker #{tracker_ind}: "
                f"{tracker['recipe']} past last step."
```